agent/connect/ca/provider_aws.go

```diff
@@ -0,0 +1,338 @@
+package ca
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/acmpca"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+type AWSProvider struct {
+	config    *structs.AWSCAProviderConfig
+	session   *session.Session
+	client    *acmpca.ACMPCA
+	isRoot    bool
+	clusterId string
+	logger    *log.Logger
+	rootPCA   *AmazonPCA
+	subPCA    *AmazonPCA
+	sleepTime time.Duration
+}
+
+func (a *AWSProvider) SetLogger(l *log.Logger) {
+	a.logger = l
+}
+
+func (a *AWSProvider) Printf(format string, v ...interface{}) {
+	if a.logger != nil {
+		a.logger.Printf(format, a)
+	}
+}
+
+func (a *AWSProvider) Configure(clusterId string, isRoot bool, rawConfig map[string]interface{}) error {
+	config, err := ParseAWSCAConfig(rawConfig)
+	if err != nil {
+		return err
+	}
+
+	sleepTime, err := time.ParseDuration(config.SleepTime)
+	if err != nil {
+		return fmt.Errorf("invalid sleep time specified: %s", err)
+	}
+	if sleepTime.Seconds() < 1 {
+		return fmt.Errorf("invalid sleep time specified: must be at least 1s")
+	}
+
+	creds := credentials.NewEnvCredentials()
+	if config.AccessKeyId != "" && config.SecretAccessKey != "" && config.Region != "" {
+		creds = credentials.NewStaticCredentials(config.AccessKeyId, config.SecretAccessKey, "")
```

Comment:
It might be worth seeing if we can have the user specify an IAM role they want along with the option to use a hardcoded access key id/secret key and then having the sdk auth using that role from the ec2 instance metadata. Just using whatever IAM role the host has access to wouldn't work, since we want to store the auth method centrally in the config here, but if we let them specify an IAM role then it should be easy for them to make sure all their servers have it. There's an example of combining multiple credential types when creating a client in the go-discover provider: https://github.com/hashicorp/go-discover/blob/master/provider/aws/aws_discover.go#L84-L95

Reply:
👍 I mentioned before using the default setup which I believe allows ENV and node-meta IAM roles too.

Reply:
One challenge here is that there's nothing to prevent you from using PCA without actually running on an EC2 instance. I guess we could allow it as an option.

Reply:
Yeah for sure we should allow this too, but I think the best would be to allow any of: creds in config file, creds in ENV, creds in default AWS config (e.g. `~/home/.awscreds`) and ec2 instance roles. I think there is a mode (and I thought it was default) for the AWS SDK to support that so most tooling uses that etc.